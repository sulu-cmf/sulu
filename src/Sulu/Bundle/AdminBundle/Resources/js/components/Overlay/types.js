// @flow
export type Action = {
    title: string,
    onClick: () => void,
};

export type Size = 'small' | 'large';
