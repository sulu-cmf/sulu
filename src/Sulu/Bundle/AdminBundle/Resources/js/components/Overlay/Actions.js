// @flow
import React from 'react';
import type {Action} from './types';
import actionsStyles from './actions.scss';

type Props = {
    actions: Array<Action>,
};

export default class Actions extends React.PureComponent<Props> {
    render() {
        const {actions} = this.props;
        if (!actions.length) {
            return null;
        }

        return (
            <div className={actionsStyles.actions}>
                {actions.map((action, index) => {
                    const handleButtonClick = action.onClick;
                    return (
                        <button
                            key={index}
                            className={actionsStyles.action}
                            onClick={handleButtonClick}
                        >
                            {action.title}
                        </button>
                    );
                })}
            </div>
        );
    }
}
