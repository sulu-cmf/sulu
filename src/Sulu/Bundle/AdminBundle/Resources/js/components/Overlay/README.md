The `Overlay` component let's you display some content above everything else.
It renders depending on the passed property and request being closed through a callback.

```
initialState = {open: false};
const actions = [
    {title: 'Destroy world', onClick: () => {/* destroy world */}},
    {title: 'Save world', onClick: () => {/* save world */}},
];
const onConfirm = () => {
    /* do confirm things */
    setState({open: false});
};

<div>
    <button onClick={() => setState({open: true})}>Open overlay</button>
    <Overlay
        title="Njan Njan Njan"
        onClose={() => setState({open: false})}
        actions={actions}
        confirmText="Apply"
        onConfirm={onConfirm}
        open={state.open}>
        <div style={{width: '900px', height: '500px', display: 'flex', alignItems: 'center', justifyContent: 'center'}}>
            <img src="http://www.nyan.cat/cats/original.gif" />
        </div>
    </Overlay>
</div>
```
