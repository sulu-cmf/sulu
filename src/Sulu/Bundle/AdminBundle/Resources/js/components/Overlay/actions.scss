@import '../../containers/Application/colors.scss';

$height: 90px;

.actions {
    line-height: $height;
}
