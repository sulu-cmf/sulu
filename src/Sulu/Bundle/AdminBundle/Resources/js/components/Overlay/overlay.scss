@import '../../containers/Application/colors.scss';
@import './variables.scss';

$primaryBackground: $concrete;
$secondaryBackground: $white;
$borderRadius: 3px;
$footerBorderColor: $silver;

$titleColor: $black;

$iconFontSize: 14px;

$confirmBackground: $shakespeare;
$confirmColor: $white;

$transitionDuration: 300ms;

.container {
    position: absolute;
    top: 0;
    left: 0;
    z-index: 100;
    width: 100%;
    height: 100%;
    display: flex;
    justify-content: center;

    & > * {
        opacity: 0;
        will-change: opacity;
        transition: opacity $transitionDuration;
    }
}

.overlay {
    position: relative;
    top: -100%;
    opacity: 0;
    will-change: top, opacity;
    transition: top $transitionDuration, opacity $transitionDuration;
    pointer-events: none;
    max-height: calc(100vh - $overlayMarginBottom);

    & > * {
        pointer-events: auto;
    }

    &.small {
        width: 50vw;
    }

    &.large {
        width: 80vw;
    }
}

.is-down {
    & > * {
        opacity: 1;
    }

    .overlay {
        top: 0;
    }
}

.content {
    border-radius: 0 0 $borderRadius $borderRadius;
    overflow: hidden;
    box-shadow: 0 10px 18px 0 rgba($mineShaft, .5);

    header {
        background: $primaryBackground;
        height: $overlayHeaderHeight;
        color: $titleColor;
        font-size: 22px;
        font-weight: bold;
        text-align: center;
        line-height: $overlayHeaderHeight;
    }

    article {
        max-height: $overlayContentMaxHeight;
        overflow: auto;
        background: $primaryBackground;
    }

    footer {
        background: $secondaryBackground;
        border-top: 1px solid $footerBorderColor;
        height: $overlayFooterHeight;
        padding: 0 60px;
        display: flex;
        justify-content: space-between;
        align-items: center;

        button:only-child {
            margin-left: auto;
        }
    }

    .icon {
        position: absolute;
        right: 60px;
        font-size: $iconFontSize;
        font-weight: bold;
        line-height: $overlayHeaderHeight;
        text-align: center;
        cursor: pointer;
    }
}
