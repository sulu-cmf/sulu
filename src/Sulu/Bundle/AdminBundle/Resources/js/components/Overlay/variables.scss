$overlayHeaderHeight: 82px;
$overlayFooterHeight: 90px;
$overlayMarginBottom: 100px;
$overlayContentMaxHeight: calc(100vh - $overlayHeaderHeight - $overlayFooterHeight - $overlayMarginBottom);
