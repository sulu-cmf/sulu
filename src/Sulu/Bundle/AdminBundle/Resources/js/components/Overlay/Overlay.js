// @flow
import classNames from 'classnames';
import Mousetrap from 'mousetrap';
import {observable, action} from 'mobx';
import {observer} from 'mobx-react';
import type {Node} from 'react';
import React from 'react';
import Portal from 'react-portal';
import Icon from '../Icon';
import Button from '../Button';
import {afterElementsRendered} from '../../services/DOM';
import Backdrop from '../Backdrop';
import type {Action} from './types';
import Actions from './Actions';
import overlayStyles from './overlay.scss';

type Props = {
    title: string,
    children: Node,
    actions: Array<Action>,
    confirmText: string,
    confirmLoading: boolean,
    onConfirm: () => void,
    open: boolean,
    onClose: () => void,
};

const CLOSE_ICON = 'times';

@observer
export default class Overlay extends React.PureComponent<Props> {
    static defaultProps = {
        open: false,
        actions: [],
        confirmLoading: false,
    };

    @observable visible: boolean = false;
    @observable openHasChanged: boolean = false;

    @action componentWillMount() {
        Mousetrap.bind('esc', this.close);
        this.openHasChanged = this.props.open;
    }

    componentWillUnmount() {
        Mousetrap.unbind('esc', this.close);
    }

    componentDidMount() {
        this.toggle();
    }

    @action componentWillReceiveProps(newProps: Props) {
        this.openHasChanged = newProps.open !== this.props.open;
    }

    componentDidUpdate() {
        this.toggle();
    }

    close = () => {
        this.props.onClose();
    };

    @action toggle() {
        afterElementsRendered(action(() => {
            if (this.openHasChanged) {
                this.visible = this.props.open;
            }
        }));
    }

    @action handleTransitionEnd = () => {
        afterElementsRendered(action(() => {
            this.openHasChanged = false;
        }));
    };

    handleIconClick = () => {
        this.close();
    };

    render() {
        const {
            open,
            title,
            actions,
            onClose,
            children,
            onConfirm,
            confirmText,
            confirmLoading,
        } = this.props;
        const containerClass = classNames(
            overlayStyles.container,
            {
                [overlayStyles.isDown]: this.visible,
            }
        );

        return (
            <div>
                <Backdrop open={open || this.openHasChanged} onClick={onClose} />
                <Portal isOpened={open || this.openHasChanged}>
                    <div
                        className={containerClass}
                        onTransitionEnd={this.handleTransitionEnd}
                    >
                        <div className={overlayStyles.overlay}>
                            <section className={overlayStyles.content}>
                                <header>
                                    {title}
                                    <Icon
                                        name={CLOSE_ICON}
                                        className={overlayStyles.icon}
                                        onClick={this.handleIconClick}
                                    />
                                </header>
                                <article>{children}</article>
                                <footer>
                                    <Actions actions={actions} />
                                    <Button skin="primary" onClick={onConfirm} loading={confirmLoading}>
                                        {confirmText}
                                    </Button>
                                </footer>
                            </section>
                        </div>
                    </div>
                </Portal>
            </div>
        );
    }
}
