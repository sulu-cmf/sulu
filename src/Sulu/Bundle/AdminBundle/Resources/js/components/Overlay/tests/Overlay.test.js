/* eslint-disable flowtype/require-valid-file-annotation */
import {mount, shallow} from 'enzyme';
import Mousetrap from 'mousetrap';
import React from 'react';
import pretty from 'pretty';
import Overlay from '../Overlay';

afterEach(() => document.body.innerHTML = '');

test('The component should render in body when open', () => {
    const body = document.body;
    const onClose = () => {};
    const view = mount(
        <Overlay
            title="My overlay title"
            onClose={onClose}
            confirmText="Apply"
            open={true}
        >
            <p>My overlay content</p>
        </Overlay>
    ).render();

    expect(view).toMatchSnapshot();
    expect(pretty(body.innerHTML)).toMatchSnapshot();
});

test('The component should render in body with actions when open', () => {
    const actions = [
        {title: 'Action 1', onClick: () => {}},
        {title: 'Action 2', onClick: () => {}},
    ];
    const body = document.body;
    const onClose = () => {};
    const view = mount(
        <Overlay
            title="My overlay title"
            onClose={onClose}
            confirmText="Apply"
            actions={actions}
            open={true}
        >
            <p>My overlay content</p>
        </Overlay>
    ).render();

    expect(view).toMatchSnapshot();
    expect(pretty(body.innerHTML)).toMatchSnapshot();
});

test('The component should not render in body when closed', () => {
    const body = document.body;
    const onClose = () => {};
    const view = mount(
        <Overlay
            title="My overlay title"
            onClose={onClose}
            confirmText="Apply"
            open={false}
        >
            <p>My overlay content</p>
        </Overlay>
    ).render();
    expect(view).toMatchSnapshot();
    expect(body.innerHTML).toBe('');
});

test('The component should request to be closed on click on backdrop', () => {
    const closeSpy = jest.fn();
    const view = shallow(
        <Overlay
            title="My overlay title"
            onClose={closeSpy}
            confirmText="Apply"
            open={true}
        >
            <p>My overlay content</p>
        </Overlay>
    );
    const backdrop = view.find('Backdrop');
    expect(backdrop.length).toBe(1);

    expect(closeSpy).not.toBeCalled();
    backdrop.props().onClick();
    expect(closeSpy).toBeCalled();
});

test('The component should request to be closed when the close icon is clicked', () => {
    const closeSpy = jest.fn();
    const view = shallow(
        <Overlay
            title="My overlay title"
            onClose={closeSpy}
            confirmText="Apply"
            open={true}
        >
            <p>My overlay content</p>
        </Overlay>
    );

    expect(closeSpy).not.toBeCalled();
    view.find('Icon').simulate('click');
    expect(closeSpy).toBeCalled();
});

test('The component should request to be closed when the esc key is pressed', () => {
    const closeSpy = jest.fn();
    mount(
        <Overlay
            title="My overlay title"
            onClose={closeSpy}
            confirmText="Apply"
            open={true}
        >
            <p>My overlay content</p>
        </Overlay>
    );

    expect(closeSpy).not.toBeCalled();
    Mousetrap.trigger('esc');
    expect(closeSpy).toBeCalled();
});

test('The component should call the callback when the confirm button is clicked', () => {
    const onClose = () => {};
    const onConfirm = jest.fn();
    const view = shallow(
        <Overlay
            title="My title"
            onClose={onClose}
            onConfirm={onConfirm}
            confirmText="Alright mate!"
        >
            <p>My overlay content</p>
        </Overlay>
    );

    expect(onConfirm).not.toBeCalled();
    view.find('Button').simulate('click');
    expect(onConfirm).toBeCalled();
});
