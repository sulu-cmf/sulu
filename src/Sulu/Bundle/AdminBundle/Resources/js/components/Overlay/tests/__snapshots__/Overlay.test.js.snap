// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The component should not render in body when closed 1`] = `<div />`;

exports[`The component should render in body when open 1`] = `<div />`;

exports[`The component should render in body when open 2`] = `
"<div>
  <div data-reactroot=\\"\\" class=\\"backdrop visible\\"></div>
</div>
<div>
  <div data-reactroot=\\"\\" class=\\"container\\">
    <div class=\\"overlay\\">
      <section class=\\"content\\">
        <header>
          <!-- react-text: 5 -->My overlay title
          <!-- /react-text --><span class=\\"icon fa fa-times\\" aria-label=\\"times\\" role=\\"button\\" tabindex=\\"0\\"></span></header>
        <article>
          <p>My overlay content</p>
        </article>
        <footer>
          <!-- react-empty: 10 -->
          <button class=\\"button primary\\"><span class=\\"text\\">Apply</span></button>
        </footer>
      </section>
    </div>
  </div>
</div>"
`;

exports[`The component should render in body with actions when open 1`] = `<div />`;

exports[`The component should render in body with actions when open 2`] = `
"<div>
  <div data-reactroot=\\"\\" class=\\"backdrop visible\\"></div>
</div>
<div>
  <div data-reactroot=\\"\\" class=\\"container\\">
    <div class=\\"overlay\\">
      <section class=\\"content\\">
        <header>
          <!-- react-text: 5 -->My overlay title
          <!-- /react-text --><span class=\\"icon fa fa-times\\" aria-label=\\"times\\" role=\\"button\\" tabindex=\\"0\\"></span></header>
        <article>
          <p>My overlay content</p>
        </article>
        <footer>
          <div class=\\"actions\\">
            <button class=\\"button link\\"><span class=\\"text\\">Action 1</span></button>
            <button class=\\"button link\\"><span class=\\"text\\">Action 2</span></button>
          </div>
          <button class=\\"button primary\\"><span class=\\"text\\">Apply</span></button>
        </footer>
      </section>
    </div>
  </div>
</div>"
`;

exports[`The component should render in body with loader instead of confirm button 1`] = `<div />`;

exports[`The component should render in body with loader instead of confirm button 2`] = `
"<div>
  <div data-reactroot=\\"\\" class=\\"backdrop visible\\"></div>
</div>
<div>
  <div data-reactroot=\\"\\" class=\\"container\\">
    <div class=\\"overlay\\">
      <section class=\\"content\\">
        <header>
          <!-- react-text: 5 -->My overlay title
          <!-- /react-text --><span class=\\"icon fa fa-times\\" aria-label=\\"times\\" role=\\"button\\" tabindex=\\"0\\"></span></header>
        <article>
          <p>My overlay content</p>
        </article>
        <footer>
          <!-- react-empty: 10 -->
          <button class=\\"button primary loading\\" disabled=\\"\\"><span class=\\"text\\">Apply</span>
            <div class=\\"loader\\">
              <div class=\\"spinner\\" style=\\"width: 25px; height: 25px;\\">
                <div class=\\"doubleBounce1\\"></div>
                <div class=\\"doubleBounce2\\"></div>
              </div>
            </div>
          </button>
        </footer>
      </section>
    </div>
  </div>
</div>"
`;
